command/agent/acl_endpoint.go

```diff
@@ -219,27 +219,25 @@ func (s *HTTPServer) aclTokenQuery(resp http.ResponseWriter, req *http.Request,
 	return out.Token, nil
 }
 
-func (s *HTTPServer) aclTokenSelf(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+func (s *HTTPServer) aclTokenSelf(resp http.ResponseWriter, req *http.Request) (any, error) {
 	if req.Method != http.MethodGet {
 		return nil, CodedError(405, ErrInvalidMethod)
 	}
-	args := structs.ResolveACLTokenRequest{}
+	args := structs.GenericRequest{}
 	if s.parse(resp, req, &args.Region, &args.QueryOptions) {
 		return nil, nil
 	}
 
-	args.SecretID = args.AuthToken
-
-	var out structs.ResolveACLTokenResponse
-	if err := s.agent.RPC("ACL.ResolveToken", &args, &out); err != nil {
+	var out structs.ACLWhoAmIResponse
+	if err := s.agent.RPC("ACL.WhoAmI", &args, &out); err != nil {
 		return nil, err
 	}
 
 	setMeta(resp, &out.QueryMeta)
-	if out.Token == nil {
+	if out.Identity.ACLToken == nil {
 		return nil, CodedError(404, "ACL token not found")
 	}
-	return out.Token, nil
+	return out.Identity.ACLToken, nil
```

Comment:
We're not really solving the problem described in https://github.com/hashicorp/nomad/issues/24663 with this, as we can only get an ACL token object back and not the rest of the identity that would be populated by WI. Similarly, if there's no ACL token and just claims, we get a 404.

We probably can't fix the issue without breaking compatibility with the old HTTP endpoint, so maybe we should make this change for backcompat and have a new API endpoint for the "full" identity object with claims? Was your plan for this PR just to do the migration and then follow-up with a fix for #24663?

Reply:
> Was your plan for this PR just to do the migration and then follow-up with a fix for https://github.com/hashicorp/nomad/issues/24663?

Yup. This PR doesn't address the underlying issue. I think the best way to address the problem of WI-JWTs-as-tokens is to introduce a separate endpoint. But I'll think more about it and get back to you on that on Monday.